MDEV-32899 InnoDB is holding shared dict_sys.latch while waiting for FOREIGN  KEY child table lock on DDL

dr-m opened the pull request:
- [x] *The Jira issue number for this PR is: MDEV-32899*
## Description
InnoDB was unnecessarily holding `dict_sys.latch` while waiting for a `FOREIGN KEY` child table lock on DDL operations. Only transactional locks and metadata locks (MDL) are designed to be blocked for a long time. Holding `dict_sys.latch` during potentially long waits will seriously disturb InnoDB operations.

`lock_table_children()`: A new function to lock all child tables of a table. We will only hold `dict_sys.latch` while traversing `dict_table_t::referenced_set`. To prevent a race condition with `std::set::erase()` we will copy the pointers to the child tables to a local vector. Once we have acquired MDL and references to all child tables, we can safely release `dict_sys.latch`, wait for the locks, and finally release the references.

`dict_acquire_mdl_shared()`: A new variant that takes `MDL_context*` instead of `THD*` as a parameter.

This fixes up commit 2ca112346438611ab7800b70bea6af1fd1169308 (MDEV-26217) and commit c3c53926c467c95386ae98d61ada87294bd61478 (MDEV-26554).
## How can this PR be tested?
This code is covered by some existing regression tests that make use of `FOREIGN KEY` constraints. Additional stress testing with RQG will be needed, including testing with the grammar that reproduced [MDEV-33104](https://jira.mariadb.org/browse/MDEV-33104).
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch.*
- [x] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced.*
## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=3021) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=3021) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=3021) it.</sub>

dr-m commented:
@mariadb-DebarunBanerjee As we discussed today, I would like to see your alternative solution to this problem. Yes, we could handle the two cases of the Boolean parameter `foreign_key_checks` separately. The key should be to demonstrate that all the code is being covered by tests.

In [MDEV-26217](https://jira.mariadb.org/browse/MDEV-26217) and in the test `innodb.foreign_key` (ever since 2ca112346438611ab7800b70bea6af1fd1169308) you can find one example of a failure that we are trying to prevent. In [MDEV-26554](https://jira.mariadb.org/browse/MDEV-26554) and  c3c53926c467c95386ae98d61ada87294bd61478 another test was added to `innodb.foreign_key`. Both involve attempts to execute concurrent DML and DDL in the `foreign_key_checks=OFF` case.

dr-m commented:
@mariadb-DebarunBanerjee I think that this update addresses most of your comments. 1cc4f98b271d33873a6257995705ca0c80e9a4e5 introduces `srw_lock_debug` instrumentation, and f50940ee0b81b9c963bd114c54788e515220bc7e will report constraint errors on `ha_innobase::delete_table()` without having to wait for any locks. There was nothing to improve on `ha_innobase::truncate()` (other than debug checks, which I added). Any `ALTER TABLE…FORCE` (no matter of `foreign_key_checks` or `ALGORITHM`) must acquire locks on the child tables so that the cache can be safely updated.
 
I did not change `ha_innobase::rename_table()` because it is being invoked also as part of `ALTER TABLE`. I did experiment with the following change, but it would cause trouble with the MDEV-18923 test in `innodb.innodb-truncate`:
```diff
@@ -14127,13 +14172,28 @@ ha_innobase::rename_table(
 
 	if (from_temp) {
 		/* There is no need to lock any FOREIGN KEY child tables. */
-	} else if (dict_table_t *table = dict_table_open_on_name(
-		    norm_from, false, DICT_ERR_IGNORE_FK_NOKEY)) {
-		error = lock_table_children(table, trx);
-		if (error == DB_SUCCESS) {
-			error = lock_table_for_trx(table, trx, LOCK_X);
+	} else {
+		const bool to_temp = dict_table_t::is_temporary_name(norm_to);
+		dict_sys.lock(SRW_LOCK_CALL);
+		if (dict_table_t *table
+		    = dict_table_open_on_name(norm_from, true,
+					      DICT_ERR_IGNORE_FK_NOKEY)) {
+			const bool check_fk = trx->check_foreigns;
+			const bool fk_error = check_fk && !to_temp
+				&& !table->referenced_set.empty();
+			dict_sys.unlock();
+			if (!check_fk) {
+				error = lock_table_children(table, trx);
+			} else if (fk_error) {
+				error = DB_CANNOT_ADD_CONSTRAINT;
+			}
+			if (error == DB_SUCCESS) {
+				error = lock_table_for_trx(table, trx, LOCK_X);
+			}
+			table->release();
+		} else {
+			dict_sys.unlock();
 		}
-		table->release();
 	}
 
 	if (strcmp(norm_from, TABLE_STATS_NAME)
```

mariadb-DebarunBanerjee reviewed (approved):
Thanks Marko. I think the comments in https://github.com/MariaDB/server/pull/2877 is addressed.
